src/import-utils.ts

```diff
@@ -0,0 +1,301 @@
+import constants from './constants';
+import * as astUtils from './ast-utils';
+import * as matchers from './matchers';
+import * as urlUtils from './url-utils';
+import {Bundle, BundleStrategy, AssignedBundle, generateBundles, BundleUrlMapper, BundleManifest, sharedBundleUrlMapper, generateSharedDepsMergeStrategy} from './bundle-manifest';
+import {BundledDocument, DocumentCollection} from './document-collection';
+import {buildDepsIndex} from './deps-index';
+import {UrlString} from './url-utils';
+
+// TODO(usergenic): Want to figure out a way to get rid of the basePath param
+// used in the inline functions, because it feels obnoxious to have to pass it
+// around for an only-occasionally used case.
+
+/**
+ * Inline the contents of the html document returned by the link tag's href
+ * at the location of the link tag and then remove the link tag.
+ */
+export async function inlineHtmlImport(
+    basePath: string|undefined,
+    docUrl: string,
+    htmlImport: ASTNode,
+    reachedImports: Set<string>,
+    docBundle: AssignedBundle,
+    manifest: BundleManifest,
+    loader: (url: UrlString) => Promise<string>) {
+  const rawImportUrl: string = dom5.getAttribute(htmlImport, 'href')!;
+  const resolvedImportUrl: string = urlLib.resolve(docUrl, rawImportUrl);
+  const importBundleUrl = manifest.bundleUrlForFile.get(resolvedImportUrl);
+
+  // Don't reprocess the same file again.
+  if (reachedImports.has(resolvedImportUrl)) {
+    astUtils.removeElementAndNewline(htmlImport);
+    return;
+  }
+
+  // If we can't find a bundle for the referenced import, record that we've
+  // processed it, but don't remove the import link.  Browser will handle it.
+  if (!importBundleUrl) {
+    reachedImports.add(resolvedImportUrl);
+    return;
+  }
+
+  // Don't inline an import into itself.
+  if (docUrl === resolvedImportUrl) {
+    reachedImports.add(resolvedImportUrl);
+    astUtils.removeElementAndNewline(htmlImport);
+    return;
+  }
+
+  // Guard against inlining a import we've already processed.
+  if (reachedImports.has(importBundleUrl)) {
+    astUtils.removeElementAndNewline(htmlImport);
+    return;
+  }
+
+  // If the html import refers to a file which is bundled and has a different
+  // url, then lets just rewrite the href to point to the bundle url.
+  if (importBundleUrl !== docBundle.url) {
+    const relative =
+        urlUtils.relativeUrl(docUrl, importBundleUrl) || importBundleUrl;
+    dom5.setAttribute(htmlImport, 'href', relative);
+    reachedImports.add(importBundleUrl);
+    return;
+  }
+
+  const document =
+      dom5.nodeWalkAncestors(htmlImport, (node) => !node.parentNode)!;
+  const body = dom5.query(document, matchers.body)!;
+  const importSource = await loader(resolvedImportUrl).catch(err => {
+    throw new Error(`Unable to load ${resolvedImportUrl}: ${err.message}`);
+  });
+
+  // Is there a better way to get what we want other than using
+  // parseFragment?
```

Comment:
That was an old comment from when AJ and I were pairing.  I got rid of it.  You answered the question anyways :)  Thanks!